cmd/fever/cmds/alertify.go

```diff
@@ -0,0 +1,170 @@
+package cmd
+
+// DCSO FEVER
+// Copyright (c) 2020, DCSO GmbH
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/DCSO/fever/input"
+	"github.com/DCSO/fever/types"
+	"github.com/DCSO/fever/util"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/buger/jsonparser"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var count uint64
+
+type alertifyAlertJSONProvider struct{}
+
+// GetAlertJSON returns the "alert" subobject for an alert EVE event.
+func (a alertifyAlertJSONProvider) GetAlertJSON(inputEvent types.Entry,
+	prefix string, ioc string) ([]byte, error) {
+	return util.GenericGetAlertObjForIoc(inputEvent, prefix, ioc,
+		"%s Generic IoC match for: %s")
+}
+
+func makeAlertifyAlertifier(prefix string) *util.Alertifier {
+	a := util.MakeAlertifier(prefix)
+	a.RegisterMatchType("dns-req", util.AlertJSONProviderDNSReq{})
+	a.RegisterMatchType("dns-resp", util.AlertJSONProviderDNSResp{})
+	a.RegisterMatchType("tls-sni", util.AlertJSONProviderTLSSni{})
+	a.RegisterMatchType("http-host", util.AlertJSONProviderHTTPHost{})
+	a.RegisterMatchType("http-url", util.AlertJSONProviderHTTPURL{})
+	a.RegisterMatchType("generic", alertifyAlertJSONProvider{})
+	a.SetExtraModifier(func(inputAlert *types.Entry, ioc string) error {
+		iocEscaped, err := util.EscapeJSON(ioc)
+		if err != nil {
+			return err
+		}
+		val, err := jsonparser.Set([]byte(inputAlert.JSONLine), iocEscaped,
+			"_extra", "vast-ioc")
+		if err != nil {
+			return err
+		}
+		inputAlert.JSONLine = string(val)
+		return nil
+	})
+	return a
+}
+
+func emitAlertsForEvent(a *util.Alertifier, e types.Entry, ioc string,
+	out io.Writer, limit uint64) error {
+	var err error
+	var alert *types.Entry
+	var specificMatch = false
+	if e.TLSSni == ioc {
+		specificMatch = true
+		alert, err = a.MakeAlert(e, ioc, "tls-sni")
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(out, "%s\n", string(alert.JSONLine))
+		if limit > 0 && count >= limit {
```

Comment:
> This `fmt.Fprintf` is repeated, so I'm wondering why not doing it at the end of the function, since you are updating `alert` and `err` in the parent scope.

I am writing each alert JSON to the `io.Writer` named `out` which in this tool prints to stdout. This could be more than one per input alert, so I need to do repeated `Fprintf`s, on per new alert. The temporary variable `alert` is overwritten each time, and I am  -- in this context -- only interested in its `JSONLine` property, which is set in `MakeAlert` and output immediately.

> count is a global which is, apparently, only used in this function? Is there a reason not putting it in this function?

It needs to keep state across multiple calls of `emitAlertsForEvent()` as it meant to be a global limit.

> And an IoC ever match more than one entry attribute? Or can it just be a switch (or else-if) here?

Yes, it can. So we might want multiple alerts per input event. I am just mirroring what the Bloom filter behaviour is right now.